src/init.cpp

```diff
@@ -797,6 +797,19 @@ ServiceFlags nLocalServices = NODE_NETWORK;
 
 }
 
+[[noreturn]] static void new_handler_terminate()
+{
+    // Rather than throwing std::bad-alloc if allocation fails, terminate
+    // immediately to (try to) avoid chain corruption.
+    // Since LogPrintf may itself allocate memory, set the handler directly
+    // to terminate first.
+    std::set_new_handler(std::terminate);
+    LogPrintf("Error: Out of memory. Terminating.\n");
```

Comment:
Before this attempt, you could do `fputs(stderr, "Error: Out of memory. Terminating.\n");` which, I believe, does not call `malloc`. At least it didn't when I wrote my own `malloc` a few years ago, but it may be different per platform. A nice to have for when a node goes down with no log entries about why.